### PR TITLE
[GUI] Support using uint8 images in fast_gui

### LINKED ISSUE #2033: In fast_gui using uint8 vector field is not displayed correctly
image
image
[ev.zip](https://github.com/taichi-dev/taichi/files/5505820/ev.zip)
Change fast_gui to false and the display will be normal.

### PR DESCRIPTION
**Pull request by archibate**
Related issue = fix #2033

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----
done in classroom with juicessh, didnt test.

### CODE REVIEW

**Comment by archibate** [comment]
Hello?

**Review by JYLeeLYJ — APPROVED** [review]
Hi! Sorry for my late attention .
This code looks great to me.
And I have test it with ```fast_gui =True``` and ```fast_gui = False```. 
Images did look the same.

File: python/taichi/lang/meta.py [hunk 1]
```diff
@@ -19,7 +19,13 @@ def tensor_to_ext_arr(tensor: ti.template(), arr: ti.ext_arr()):
 def vector_to_fast_image(img: ti.template(), out: ti.ext_arr()):
     # FIXME: Why is ``for i, j in img:`` slower than:
     for i, j in ti.ndrange(*img.shape):
-        r, g, b = min(255, max(0, int(img[i, img.shape[1] - 1 - j] * 255)))
+        r, g, b = 0, 0, 0
+        clr = img[i, img.shape[1] - 1 - j]
```
Comment (yuanming-hu) [inline python/taichi/lang/meta.py:23]: What does `clr` mean?

**Review by yuanming-hu — COMMENTED** [review]
Cool! My only concern is the `clr` variable name.

**Comment by ffcr** [comment]
In addition, when arch is OpenGL, it will crash and display "Type u8 not supported."

**Comment by archibate** [comment]
> In addition, when arch is OpenGL, it will crash and display "Type u8 not supported."

Yes, this is expected, uint8 is not supported by OpenGL compute shader.

File: taichi/codegen/codegen_llvm.cpp [hunk 2]
```diff
@@ -1231,20 +1233,31 @@ void CodeGenLLVM::visit(LinearizeStmt *stmt) {
...
-    emit(R"({}* {}[1] {{({} *)((char *){}[0] + {})}};)", dtn, stmt->raw_name(),
-         dtn, stmt->input->raw_name(), stmt->offset);
-  } else {
-    emit(R"(auto {} = {} + {};)", stmt->raw_name(), stmt->input->raw_name(),
-         stmt->offset);
-  }
-  */
+void CodeGenLLVM::visit(IntegerOffsetStmt *stmt){TI_NOT_IMPLEMENTED}
+
+llvm::Value *CodeGenLLVM::create_bit_ptr_struct(llvm::Value *byte_ptr_base,
```
Comment (k-ye) [inline taichi/codegen/codegen_llvm.cpp:1238]: Are these bit types changes related to fast_gui?

Comment (archibate) [reply taichi/codegen/codegen_llvm.cpp:1238]: Mismerge, sorry!

File: python/taichi/lang/meta.py [hunk 3]
```diff
@@ -19,7 +19,13 @@ def tensor_to_ext_arr(tensor: ti.template(), arr: ti.ext_arr()):
...
     # FIXME: Why is ``for i, j in img:`` slower than:
     for i, j in ti.ndrange(*img.shape):
-        r, g, b = min(255, max(0, int(img[i, img.shape[1] - 1 - j] * 255)))
+        r, g, b = 0, 0, 0
+        color = img[i, img.shape[1] - 1 - j]
+        if ti.static(img.dtype in [ti.f32, ti.f64]):
+            r, g, b = min(255, max(0, int(color * 255)))
+        else:
+            ti.static_assert(img.dtype == ti.u8)
+            r, g, b = clr
```
Comment (k-ye) [inline python/taichi/lang/meta.py:28]: `color`?

**Review by k-ye — APPROVED** [review]
lgtm

**Pull request merged**